### PR TITLE
[DRAFT] `expect` function for testing

### PR DESCRIPTION
**Pull request by Sorikairox**
Draft to get feedback on implementing them as wrappers for `assert` functions.

Tests for `BeEqual` matcher are exactly the same as `assertEquals`, so we might be able to refactor them into one test suit calling both instead of having duplication.

Base code for expect from https://github.com/allain/expect/tree/master which is ISC License.

Modified it to have Matchers get a context and throw as they use `assert` function. Also fixed some typing.

Related to issue #1779

### CODE REVIEW

File: expect/expect.ts [hunk 1]
```diff
@@ -0,0 +1,129 @@
+// Copyright 2018-2023 the Deno authors. All rights reserved. MIT license.
+
+//ISC License
```
Comment (kt3k) [inline expect/expect.ts:3]: I think we can use some short form of license header like:

```
Copyright 2019 Allain Lalonde. All rights reserved. ISC License.
```

Comment (Sorikairox) [reply expect/expect.ts]: Done !

**Comment by kt3k** [comment]
I think `matcher/` dir and `types.ts` should be named `_matcher/` and `_types.ts` respectively to minimize the publicly exposed APIs (I suppose the only meaningful API for users to import is `expect`).

**Comment by bartlomieju** [comment]
Pinging @marvinhagemeister since he brought up a need for this API recently :)

**Comment by marvinhagemeister** [comment]
Nothing to say other than I really like this and can't wait to have this natively in deno 👍

**Comment by bartlomieju** [comment]
> @bartlomieju @marvinhagemeister Do you have any opinion/preference about the export path of this module? This PR suggests `std/expect/expect.ts`. I think that would make sense. But The original issue suggested `std/testing/expect.ts` that also seems reasonable to me.

I would err on `std/exect/mod.ts` (similar to how `assert` was made top-level). `std/testing/expect.ts` works too 👍

**Comment by aapoalas** [comment]
One argument for why assert was made top level was that it has valid runtime uses as well. Expect kind of doesn't. I'd put it under testing.

**Comment by Sorikairox** [comment]
FWIW, I was going to put it in `std/testing/expect.ts` at first, and then I saw `assert` at the top level and moved everything. So I think `testing/expect.ts` fit better.

**Comment by marvinhagemeister** [comment]
@Sorikairox I'd really like us to re-open this PR. I've been using `npm:expect` in one project recently and it's super slow. It does a lot of allocations and collects error stack traces upfront and other things. It's super slow even if you do just `expect(1).toEqual(1)` and none of that is needed.

| assertion library | time |
|---|--:|
| `npm:expect` | 786ms |
| assert | 144ms |

Made a super tiny wrapper around `assert` to mirror the `expect` API like in this PR and the speedup is too good to ignore. I want us to ship `expect` ourselves. I really want to see this land.

**Comment by bartlomieju** [comment]
I would also be in favor of landing as is and adding additional APIs in follow up PRs.

**Comment by Sorikairox** [comment]
Hey guys. My apologies, life happened and I did not manage to make time to continue on this.

Due to rebasing and force pushing, looks like I cannot re-open this one. Opening https://github.com/denoland/deno_std/pull/3627

**Pull request closed without merging**